### ISSUE TITLE
Where is the code that can teach me to generate a tokenezed url from session?

### ISSUE THREAD

**Issue by shaenr** [ytdl-org/youtube-dl#29966]
## Checklist

- [ x] I'm asking a question
- [x ] I've looked through the README and FAQ for similar questions
- [ x] I've searched the bugtracker for similar questions including closed ones

## Question

WRITE QUESTION HERE

Here's an example command you could do. 

`youtube-dl https://vektroid.bandcamp.com/track/contemporary-from-new-dreams-ltd-demo --get-url`
This will return a url that is like `https://t4.bcbits.com/stream/d527df490e1f4a0509b28550b2ecce99/mp3-128/3872093290?p=0&ts=sdafdsaf&t=sadfadsfsdfdsafdsfdsf&token=asdfdsafadsfdsafsdfdasf` except I made up all the parameters with `sdfsdafdsf`.

I looked in the code for the `extractors/bandcamp.py` for something like that... But it seems like just tests and an expected regex.

Basically, I would love to use this code in a client side app. But I cant call youtube-dl from some client. So I have to figure out how to do the same thing in JS to get that url for a user.

**Comment by dirkf** [comment]
Use the Python debugger and/or the `--print-traffic` option to see what's happening. The URL is presumably being set in line 180. It needed the original page and two subsequent JSON fetches to retrieve the URL. On the one hand this is good news since you too can do that in client-side JS; on the other hand, it doesn't tell you how to make the URL since the extractor is just uisng the URL constructed by the site.

**Comment by 89z** [comment]
@dirkf do you really need to make nonsense replies to every post?

@shaenr, to actually answer the question, currently the program just makes a request like this:

~~~
GET /track/contemporary-from-new-dreams-ltd-demo HTTP/1.1
Host: vektroid.bandcamp.com
~~~

then in the HTML, you can find something like this:

~~~
<script type="application/ld+json">
{"inAlbum":{"@id":"https://duststoredigital.com/album/silenced","name":"Silenc...
</script>
~~~

where you can parse the text node as JSON, and get the data you need. I am trying to see if JSON only method is available, maybe using credentials from the Android app. If I find anything, I will update.

**Comment by dirkf** [comment]
I think you just restated my point that the extractor fetches the webpage and makes two further requests to get the media URL, which is therefore not, as OP appeared to hope, calculated by the extractor, even though he could replicate the logic with client-side JS. Personally, I favour teaching people to fish. But maybe everyone's talking nonsense.

**Comment by 89z** [comment]
> The URL is presumably being set in line 180.

This is a low value comment, as it will likely be invalid the next time a commit is made. In addition, you can just link to the actual source, like this:

<https://github.com/ytdl-org/youtube-dl/blob/a8035827177d6b59aca03bd717acb6a9bdd75ada/youtube_dl/extractor/bandcamp.py#L180>

> It needed the original page and two subsequent JSON fetches

If YouTube-DL is actually doing that, its pointless, as a single request to the HTML page, will get you direct media URL.

> On the one hand this is good news since you too can do that in client-side JS

Unless you actually know how to reverse engineer the JS to do that, this was a pointless comment.

> Personally, I favour teaching people to fish. But maybe everyone's talking nonsense.

Maybe instead of "information gatekeeping", if youre going to take it upon yourself to comment on every issue, actually give the information people are asking for.

**Comment by dirkf** [comment]
> .... In addition, you can just link to the actual source, like this:
> ...

As any fule kno, but sadly rarely done. One reason is that GH's permalinks aren't automatically linked to the commit that is being linked when the commit is not explicit in the URL (which would be a useful enhancement); another, as in this case, is that the reference came from a local copy of the latest source.

> Although it seems that line has nothing to do with what OP asked.

It's where the media URL, that is returned by `--get-url`, is being set, which is exactly what was asked.

> > It needed the original page and two subsequent JSON fetches
> 
> If YouTube-DL is actually [fetching the original page, then two further fetches], its pointless, as a single request to the HTML page, will get you direct media URL.

It is, and, if this is consistent across Bandcamp sites, it could instead use the `<script type="application/ld+json">` element, perhaps specialising the `_json_ld()` extractor method which doesn't support the specific ld+json structure being sent.

> > On the one hand this is good news since you too can do that in client-side JS
> 
> Unless you actually know how to reverse engineer the JS to do that, this was a pointless comment.

I'm sure that OP is capable of putting the necessary fetches into JS, since that was the preferred target. It will be even better news that it's only necessary to fetch the page and extract the ld+json object.

>...."information gatekeeping"...

How could any gate can be kept around open source code published in GitHub and elsewhere? What's needed is people who are prepared to look into and understand the code, as OP was, rather than leeching off the code-base.
